### ISSUE TITLE
Multi-Zone Node Template

### ISSUE THREAD

**Issue by JasonvanBrackel** [rancher/rancher#15739]
Currently a node template in the various IaaS providers is connected to a specific location.  
![image](https://user-images.githubusercontent.com/942110/45836213-330a3700-bcda-11e8-8d69-122ed714178d.png)

This means that the user has to clone and manage the template for every location.  

I would suggest changing the feature so that node templates are not attached to a location, but node pools are associated to a specific location.

**Comment by vincent99** [comment]
This doesn't really work.

There's no standard config field for "location".  Some providers have no equivalent, some have 1 field, some have 2.  So for a custom machine driver we'd have no idea which fields should be in vs not-in the template.  (Yes the UI has a [mapping](https://github.com/rancher/ui/blob/master/lib/shared/addon/mixins/node-driver.js#L12-L43) for common providers, but best-guess for display is very different than filling in a required field on create).

And then each provider has another set of fields that have to vary depending on the selected location.  e.g. for Amazon the subnets, AMIs, and security group names are all going to be different if you pick a different region.  Instance types that are available vary by location in many providers, etc.

**Comment by ekristen** [comment]
You could define a common mapping for R2 that maps down to the specific providers to help with this.

I agree having to maintain multiple templates for different locations is a pain. 

Another option would be to teach node pools to use multiple templates and spread the instance counts over the templates.

**Comment by JasonvanBrackel** [comment]
@vincent99 I didn't say it was easy.  It's just not very good UX.  It doesn't have to be what I suggest, you know the code suggest a reasonable alternative to status quo.

**Comment by vincent99** [comment]
Just for Amazon, all these depend on what zone you pick:
  - VPC/Subnet
  - Security Group (depends on VPC)
  - AMI
  - SSH User (may vary on AMI)
  - Instance Type (_should_, but I don't think we currently do this)

So if you pull region/zone out of the template, now you have to ask 5 more things when trying to use the template, and I don't know what the point of a template even is at that point.

**Comment by ekristen** [comment]
Fair point, but as you said it's a template.

I would say the only thing that matters is the AZ and Subnets.

If a template is multi-zoned, all this information would be the same:

* VPC
* Security Group
* AMI
* SSH User
* Instance Type

The only thing IMO that would differ in the template would be subnet to zone mapping.

**Comment by vincent99** [comment]
That's my point, they _aren't_ the same, they either can or _have to_ vary depending on the selected zone.  Security groups and AMIs are defined per region.  Newer zones don't offer older instance types.  Etc.  For every provider, differently.

**Comment by cloudnautique** [comment]
If I understand the initial request, this is getting at what AWS does with an ASG. In AWS you have a launch config that describes the node and then define the auto-scaling group that defines where the instances with that config will run.

So, thinking outloud here, if we had a node template for AWS that only described the specifics of the instance tied to a region. Then gave config options at the node pool level that defined the subnets, vpc, qty etc. Maybe we lock a cluster down to a single VPC.

If a list of subnets/azs were passed in, Rancher could handle distributing the instances as the scale increased. Security groups at the nodepool level seem like they would be a win because you could do one per role type and not per node template.

We will need to figure out the right way to decouple these so we can get to providing Rancher based autoscaling groups.

**Comment by ekristen** [comment]
I agree on the decoupling. Right now node templates are 1 to 1 based on Docker Machine's capabilities per driver. To truly support an auto-scaling system that's driven by Rancher that works across providers you have to decouple and abstract away a Docker Machine Node Template vs a Rancher Node Template.

With respect to a security group per role type, I love it. That's what I'm doing right now but manually. I'm basically having to manage node templates per AZ and then control instance count per AZ. I've got etcd/controlplane on the same node but I have the instance count 1 per AZ.

**Comment by ekristen** [comment]
While we are on the topic of ASG capabilities, it would be nice to look at looping in capabilities like autospotting, replacing on-demand with spot instances for worker node templates, and the ability to scale up and down based on workloads both from Rancher monitoring and external API interaction (but I think this last one already is possible)

**Comment by JasonvanBrackel** [comment]
@cloudnautique Multi-zone might be the wrong term.  Really multi-location. The hard lock between the template and the node driver is requiring cloning of node templates in a way that is frustrating our users.  Putting a level of abstraction on this to make the UX friendlier (like we do with RBAC and projects), would go a long way.

**Comment by vincent99** [comment]
I really don't thinking you guys are thinking through what the usability of this would be.  Splitting the config means you're making it more "expensive" to use the template **every time** you create a pool by asking for the region/zone and several dependent options.  In order to avoid a couple **one-time** clones where you click though and change exactly the same options.

This plays poorly with other things people want, like sharing node templates, and restricting clusters to only those shared templates.  

And this greatly exacerbates the other thing where the UI needs the API keys to let the driver make API calls to show you those options.  That still has no practical solution, but now it's a problem extended to even _using_ a template or editing a cluster.

And I think it's way weirder if you take some middle ground for Amazon and say it's only for zones within a region instead of region+zone.  That's like the worst of both worlds, I still have to clone for different regions and have to pick zone-specific options per node pool anyway.

If anything, it seems like the real problem here is "I'm trying to create a pool now and there's no template for the right location, but _this one_ is almost right.  Going back and cloning one is not convenient **from here**, right now." That could be solved in far simpler ways, like a "clone existing..." option in the template drop-down.

**Comment by mohag** [comment]
If Rancher had an option to manage / create the VPC (or at least the subnets) as well, it would be much simpler... (But that is a separate feature request)

(AZ support within a node-template seems useful, multiple regions would likely complicate stuff more than needed)

KOPS ends up with instancegroups per AZ for manager (controlplane + etcd) and a cross-AZ one for workers (where it is less critical to have them exactly balanced) ([Where all the relevant subnets are listed](https://github.com/kubernetes/kops/blob/master/docs/instance_groups.md#creating-a-new-instance-group)) (You don't actually set the AZ, you just set the subnet(s)) ([This](https://github.com/kubernetes/kops/blob/master/docs/instance_groups.md#moving-from-one-instance-group-spanning-multiple-azs-to-one-instance-group-per-az) also seems relevant - why you might want to force a specific AZ at times)

**Comment by reubenavery** [comment]
+1 on this.  with EC2 this immediately defeats the purpose of Node Templates and the corresponding Node Pools.  Templates (for EC2) need to be bound only by VPC (implying Region).  Subnets ought to be selected in a multiple fashion when deploying a pool of nodes based on the template.  When I'm building for HA, I certainly don't want every node instance in my etcd/worker/control plane pool to be on the same subnet and thus same AZ.  And creating a seperate node template for each separate subnet rather defeats the purpose as this pattern becomes 1:1 for node template to node.

My two cents
